rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,427 @@
+        if self.is_released {
+            return true;
+        }
+        now_millis() - self.lookup_time > (self.lease_duration as u128) * 1000
+    }
+}
+
+fn now_millis() -> u128 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis()
+}
+
+pub fn attr<T: ToString>(s: T) -> AttributeValue {
+    AttributeValue {
+        s: Some(s.to_string()),
+        ..Default::default()
+    }
+}
+
+struct AcquireLockState<'a> {
+    client: &'a DynamoLock,
+    active_lock: Option<LockItem>,
+    started: Instant,
+    timeout_in: Duration,
+}
+
+impl<'a> AcquireLockState<'a> {
+    fn has_timed_out(&self) -> bool {
+        self.started.elapsed() > self.timeout_in
+    }
+
+    async fn try_acquire_lock(&mut self) -> Result<LockItem, DynamoError> {
+        match self.client.get_lock().await? {
+            None => {
+                // there's no lock, we good to acquire it
+                Ok(self.upsert_new_lock().await?)
+            }
+            Some(existing) if existing.is_released => {
+                // lock is released by a caller, we good to acquire it
+                Ok(self.upsert_released_lock(existing.data).await?)
+            }
+            Some(existing) => {
+                // there's existing lock and it's out first attempt to acquire it
+                if self.active_lock.is_none() {
+                    // first we store it, extend timeout period and try again later
+                    return self.set_new_active_lock_and_fail(existing);
+                }
+
+                // there's existing lock and we've already tried to acquire it, let's try again
+                let active = self.active_lock.as_ref().unwrap();
+                let active_rvn = &active.record_version_number;
+
+                // let's check store rvn against current lock from dynamo
+                if active_rvn == &existing.record_version_number {
+                    // rvn matches
+                    if active.is_expired() {
+                        // the lock is expired and we're safe to try to acquire it
+                        self.upsert_expired_lock(active_rvn, existing.data).await
+                    } else {
+                        // the lock is not yet expired, try again later
+                        Err(DynamoError::ConditionalCheckFailed)
+                    }
+                } else {
+                    // rvn doesn't match, meaning that other worker acquire it before us
+                    // let's change active lock with new one and extend timeout period
+                    self.set_new_active_lock_and_fail(existing)
+                }
+            }
+        }
+    }
+
+    fn set_new_active_lock_and_fail(&mut self, lock: LockItem) -> Result<LockItem, DynamoError> {
+        self.extend_timeout_by(lock.lease_duration);
+        self.active_lock = Some(lock);
+
+        Err(DynamoError::ConditionalCheckFailed)
+    }
+
+    fn extend_timeout_by(&mut self, seconds: u64) {
```

Comment:
nitpick, this function is only used in `set_new_active_lock_and_fail`. unless you are planning to use it in other places, it would have been better to just unroll and inline it within that function, in general less indirection makes the code easier to follow.

Reply:
yeah, it was used previously in two places but no anymore so that's makes total sense to inline it